contracts/liquidityMining/KyberFairLaunch.sol

```diff
@@ -0,0 +1,469 @@
+  */
+  function renewPool(
+    uint256 _pid,
+    uint32 _startBlock,
+    uint32 _endBlock,
+    uint32 _rewardLockBps,
+    uint128 _rewardPerBlock,
+    bool _withUpdate
+  ) external override onlyAdmin {
+    require(_pid < poolInfo.length, 'renew: invalid pool id');
+    // only update this pool
+    if (_withUpdate) updatePoolRewards(_pid);
+
+    PoolInfo storage pool = poolInfo[_pid];
+    {
+      // prevent stack too deep
+      uint256 currentBlock = block.number;
+      require(
+        pool.startBlock > currentBlock || pool.endBlock < currentBlock,
+        'renew: pool is active'
+      );
+
+      require(
+        _startBlock > currentBlock && _endBlock > _startBlock, 'renew: invalid blocks'
+      );
+      require(_rewardLockBps <= BPS, 'renew: invalid lock bps');
+      require(_rewardPerBlock > 0, 'renew: invalid reward per block');
+    }
+
+    {
+      // prevent stack too deep
+      (
+        pool.startBlock,
+        pool.endBlock,
+        pool.rewardLockBps,
+        pool.lastRewardBlock
+      ) = (
+        _startBlock,
+        _endBlock,
+        _rewardLockBps,
+        _startBlock
+      );
+    }
+
+    (pool.rewardPerBlock, pool.accRewardPerShare) = (_rewardPerBlock, 0);
+
+    emit RenewPool(_pid, _startBlock, _endBlock, _rewardLockBps, _rewardPerBlock);
+  }
+
+  /**
+  * @dev Update a pool, allow to change end block, reward per block and lock bps
+  * @param _pid: pool id to be renew
+  * @param _endBlock: block where the reward ends
+  * @param _rewardLockBps: percentage (in bps) of reward to be locked
+  * @param _rewardPerBlock: amount of reward token per block for the pool
+  * @param _withUpdate: whether to update the pool
+  */
+  function updatePool(
+    uint256 _pid,
+    uint32 _endBlock,
+    uint32 _rewardLockBps,
+    uint128 _rewardPerBlock,
+    bool _withUpdate
+  ) external override onlyAdmin {
+    require(_pid < poolInfo.length, 'update: invalid pool id');
+    // only update this pool
+    if (_withUpdate) updatePoolRewards(_pid);
+
+    PoolInfo storage pool = poolInfo[_pid];
+
+    require(_endBlock > block.number && _endBlock > pool.startBlock, 'update: invalid end block');
+    require(_rewardLockBps <= BPS, 'update: invalid lock bps');
+    require(_rewardPerBlock > 0, 'update: invalid reward per block');
+
+    (
+      pool.endBlock,
+      pool.rewardLockBps,
+      pool.rewardPerBlock
+    ) = (
+      _endBlock,
+      _rewardLockBps,
+      _rewardPerBlock
+    );
+    emit UpdatePool(_pid, _endBlock, _rewardLockBps, _rewardPerBlock);
+  }
+
+  /**
+  * @dev deposit to tokens to accumulate rewards
+  * @param _pid: id of the pool
+  * @param _amount: amount of stakeToken to be deposited
+  */
+  function deposit(uint256 _pid, uint256 _amount) external override {
+    PoolInfo storage pool = poolInfo[_pid];
+    UserInfo storage user = userInfo[_pid][msg.sender];
+    require(pool.stakeToken != address(0), 'deposit: not accept deposit');
```

Comment:
this  is redundant because stateToken is different from address(0) when addNewPool

Reply:
if we remove this check, we will need to check for pid in range, i.e < poolLength